pkg/backup/backup_compaction.go

```diff
@@ -92,128 +95,273 @@ func StartCompactionJob(
-		execCtx.User(), false,
-	)
-	if err != nil {
+	if backupManifest == nil || forceReadBackupManifest {
+		backupManifest, memSize, err = b.readManifestOnResume(ctx, &mem, p.ExecCfg(), defaultStore,
+			updatedDetails, p.User(), &kmsEnv)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We retry on pretty generic failures -- any rpc error. If a worker node were
+	// to restart, it would produce this kind of error, but there may be other
+	// errors that are also rpc errors. Don't retry too aggressively.
+	retryOpts := retry.Options{
+		MaxBackoff: 1 * time.Second,
+		MaxRetries: 5,
+	}
+
+	if p.ExecCfg().BackupRestoreTestingKnobs != nil &&
+		p.ExecCfg().BackupRestoreTestingKnobs.BackupDistSQLRetryPolicy != nil {
+		retryOpts = *p.ExecCfg().BackupRestoreTestingKnobs.BackupDistSQLRetryPolicy
+	}
+
+	if err := p.ExecCfg().JobRegistry.CheckPausepoint("backup.before.flow"); err != nil {
 		return err
 	}
-	defer func() {
-		mem.Shrink(ctx, memReserved)
-	}()
-	return compactIncrementals(
-		ctx, jobID, execCtx, details, manifests, encryption, &kmsEnv, localityInfo,
+
+	// We want to retry a backup if there are transient failures (i.e. worker nodes
+	// dying), so if we receive a retryable error, re-plan and retry the backup.
+	// TODO (kev-cao): Add progress tracking to compactions.
+	for r := retry.StartWithCtx(ctx, retryOpts); r.Next(); {
+		if err = compactChain.Compact(ctx, p, updatedDetails, backupManifest, defaultStore, &kmsEnv); err == nil {
+			break
+		}
+
+		if joberror.IsPermanentBulkJobError(err) {
+			return errors.Wrap(err, "failed to run backup compaction")
+		}
+
+		// If we are draining, it is unlikely we can start a
+		// new DistSQL flow. Exit with a retryable error so
+		// that another node can pick up the job.
+		if p.ExecCfg().JobRegistry.IsDraining() {
+			return jobs.MarkAsRetryJobError(errors.Wrapf(err, "job encountered retryable error on draining node"))
+		}
+
+		log.Warningf(ctx, "encountered retryable error: %+v", err)
+
+		// Reload the backup manifest to pick up any spans we may have completed on
+		// previous attempts.
+		var reloadBackupErr error
+		mem.Shrink(ctx, memSize)
+		backupManifest, memSize, reloadBackupErr = b.readManifestOnResume(ctx, &mem, p.ExecCfg(),
+			defaultStore, updatedDetails, p.User(), &kmsEnv)
+		if reloadBackupErr != nil {
+			return errors.Wrap(reloadBackupErr, "could not reload backup manifest when retrying")
+		}
+	}
+	// We have exhausted retries without getting a "PermanentBulkJobError", but
+	// something must be wrong if we keep seeing errors so give up and fail to
+	// ensure that any alerting on failures is triggered and that any subsequent
+	// schedule runs are not blocked.
+	if err != nil {
+		return errors.Wrap(err, "exhausted retries")
+	}
+
+	return b.maybeNotifyScheduledJobCompletion(
+		ctx, jobs.StateSucceeded, p.ExecCfg().JobsKnobs(), p.ExecCfg().InternalDB,
 	)
 }
 
-func compactIncrementals(
+func (b *backupResumer) OnFailOrCancelCompaction(
```

Comment:
nit: i took a closer look at the og onfailorcancel routine and it actually noops if there's no pts on the job record, so I actually think the og routine is totally fine. I think you can delete this. 

We'll worry about metrics/telemetry later.